Compression Streams

stwrt opened the pull request:
#### f044be813053dc25cb5ef857d8fde1e689e3c3df
<pre>
Compression Streams
<a href="https://bugs.webkit.org/show_bug.cgi?id=242665">https://bugs.webkit.org/show_bug.cgi?id=242665</a>

Reviewed by Brent Fulgham.

This patch implements the Compression Streams API.
Currently we are passing all WPT test cases (except for an initialization one where
we are throwing a typeError instead of a standard Error).

This implementation uses the compression and decompression functions from the Zlib library.

For compression I attempt to perform the compression in one shot by allocating enough memory up front per chunk.
For decompression I am using a scaling allocation approach. We first allocate 1 page (16KB) of memory, then two, then 4 and so on...
I added a hard limit of 1GB per allocation to make sure I don&apos;t blow away all the device&apos;s memory.
For every chunk, I take all these allocations and merge them into one array.

* LayoutTests/imported/w3c/web-platform-tests/compression/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.js: Added.
(get value):
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.js: Added.
(async compressChunkList):
(const.chunkList.of.chunkLists.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.js: Added.
(makeExpectedChunk):
(async compressMultipleChunks):
(numberOfChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.js: Added.
(async compressArrayBuffer):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.js: Added.
(async compressArrayBuffer):
(test):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-with-detach.tentative.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-with-detach.tentative.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-with-detach.tentative.window.js: Added.
(createLargeRandomInput):
(decompress):
(promise_test.async const):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.js: Added.
(get value):
(return.new.Uint8Array):
(value.new.Uint8Array):
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.js: Added.
(value.new.Int8Array.new.Uint8Array):
(value.new.Uint8Array):
(value.new.Uint8ClampedArray):
(value.new.Int16Array):
(value.new.Uint16Array):
(value.new.Int32Array):
(value.new.Uint32Array):
(value.new.Float32Array):
(value.new.Float64Array):
(const.chunk.of.bufferSourceChunksForDeflate.promise_test.async t):
(const.chunk.of.bufferSourceChunksForGzip.promise_test.async t):
(const.chunk.of.bufferSourceChunksForDeflateRaw.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.js: Added.
(test.t.new.DecompressionStream.toString):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.js: Added.
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any.js: Added.
(async tryDecompress):
(corruptInput):
(of.expectations.promise_test.async const):
(of.expectations.promise_test):
(of.expectations.of.fields.of.cases.promise_test.async const):
(of.expectations.of.fields.of.cases.promise_test):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.js: Added.
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.js: Added.
(async decompressArrayBuffer):
(chunkSize.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.js: Added.
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-with-detach.tentative.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-with-detach.tentative.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-with-detach.tentative.window.js: Added.
(async createLargeCompressedInput):
(promise_test.async const):
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness-shadowrealm.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness-shadowrealm.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness-shadowrealm.window.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/resources/concatenate-stream.js: Added.
(async concatenateStream):
* LayoutTests/imported/w3c/web-platform-tests/compression/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/third_party/pako/LICENSE: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/third_party/pako/README: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/third_party/pako/pako_inflate.min.js: Added.
(l):
(return.r):
(i.assign):
(i.shrinkBuf):
(a.arraySet):
(a.flattenChunks):
(r.arraySet):
(r.flattenChunks):
(i.setTyped):
(1):
(d):
(i.string2buf):
(i.buf2binstring):
(i.binstring2buf):
(i.buf2string):
(i.utf8border):
(2):
(t.exports):
(3):
(4):
(s):
(5):
(6):
(7):
(L):
(r):
(o):
(f):
(H):
(j):
(i.inflateInit):
(i.inflate):
(i.inflateEnd):
(i.inflateGetHeader):
(i.inflateSetDictionary):
(8):
(9):
(10):
(11):
(o.prototype.push):
(o.prototype.onData):
(o.prototype.onEnd):
(i.inflateRaw):
(string_appeared_here):
* LayoutTests/imported/w3c/web-platform-tests/compression/third_party/pako/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/w3c-import.log: Added.
* Source/WebCore/Modules/compression/CompressionStream.js:
(initializeCompressionStream):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
(WebCore::CompressionStreamEncoder::encode):
(WebCore::CompressionStreamEncoder::flush):
(WebCore::CompressionStreamEncoder::initialize):
(WebCore::CompressionStreamEncoder::compress):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
(WebCore::CompressionStreamEncoder::~CompressionStreamEncoder):
(WebCore::CompressionStreamEncoder::CompressionStreamEncoder):
* Source/WebCore/Modules/compression/DecompressionStream.js:
(initializeDecompressionStream):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::decode):
(WebCore::DecompressionStreamDecoder::flush):
(WebCore::DecompressionStreamDecoder::initialize):
(WebCore::DecompressionStreamDecoder::decompress):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
(WebCore::DecompressionStreamDecoder::~DecompressionStreamDecoder):
(WebCore::DecompressionStreamDecoder::DecompressionStreamDecoder):
* Source/WebCore/Modules/compression/Formats.h:

Canonical link: <a href="https://commits.webkit.org/253093@main">https://commits.webkit.org/253093@main</a>
</pre>

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.js:
```diff
@@ -49,12 +53,15 @@ function initializeCompressionStream(format)
             buffer = encoder.@encode(chunk);
         } catch (e) {
-            return @Promise.@reject(e);
+             return @Promise.@reject(@makeTypeError(e.message));
```
Looks like an incorrect leading space here.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStream.js:
```diff
@@ -66,7 +73,13 @@ function initializeCompressionStream(format)
+        }
         if (buffer) {
             const transformStream = @getByIdDirectPrivate(this, "CompressionStreamTransform");
             const controller = @getByIdDirectPrivate(transformStream, "controller");
```
It seems like it would be simpler just to have this entire block inside the 'try' section, rather than declaring 'buffer' outside, since it's never used again after these steps.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -26,21 +26,125 @@
 {
-    UNUSED_PARAM(input);
-    return nullptr;
+    size_t length = input.length();
```
I'm not sure this declaration is useful unless 'input.length()' doesn't return `size_t` for some reason.

stwrt replied:
Nice catch. Was a left over from a past iteration of the code.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -26,21 +26,125 @@
+ExceptionOr<Vector<uint8_t>> CompressionStreamEncoder::compress(const uint8_t* input, const size_t inputLength)
+{
+    size_t index = 0, totalSize = 0;
+    size_t allocateSize = (inputLength < 16384) ? 16384 : inputLength;
```
The magic number 16384 should be declared as a constant somewhere to explain what it is.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -26,21 +26,125 @@
+            shouldCompress = false;
+        else {
+            index++;
+            if (allocateSize < 1073741824) // 1GB
```
Again, this magic number should be declared as a meaningfully-named constant.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -26,21 +26,125 @@
+
+    Vector<uint8_t> output;
+    output.reserveCapacity(totalSize);
+    for (size_t i = 0; i <= index; i++)
```
You should use a modern 'for' loop here:

```
for (auto& storageElement : storage)
    output.append(storageElement.data(), storageEleemnt.size());
```

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -41,17 +42,31 @@ class CompressionStreamEncoder : public RefCounted<CompressionStreamEncoder> {
 private:
-    const Formats::CompressionFormat m_format;
+
+    bool initailized;
```
```
bool initialized { false };
bool finish { false };
```

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -41,17 +42,31 @@ class CompressionStreamEncoder : public RefCounted<CompressionStreamEncoder> {
     explicit CompressionStreamEncoder(unsigned char format) 
-        : m_format(static_cast<Formats::CompressionFormat>(format))
+        : initailized(false)
+        , finish(false)
```
... then. you can get rid of these two initializations.

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStream.js:
```diff
@@ -66,7 +71,13 @@ function initializeDecompressionStream(format)
+        
+        let buffer;
+        try {
+        buffer = decoder.@flush();
```
`buffer` doesn't seem to be indented here.

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStream.js:
```diff
@@ -66,7 +71,13 @@ function initializeDecompressionStream(format)
+        }
         if (buffer) {
             const transformStream = @getByIdDirectPrivate(this, "DecompressionStreamTransform");
             const controller = @getByIdDirectPrivate(transformStream, "controller");
```
Again: Propose you move this stuff inside the 'try' block, since it's never used again.

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -32,15 +32,121 @@
 {
-    UNUSED_PARAM(input);
-    return nullptr;
+    size_t length = input.length();
```
I'd just use `input.length()` later in the method, since this isn't used anywhere else.

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -32,15 +32,121 @@
+ExceptionOr<bool> DecompressionStreamDecoder::initialize() 
 {
-    return nullptr;
+    int result;
```
I'd initialize this as:

`int result = Z_OK;`

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -32,15 +32,121 @@
 
+ExceptionOr<Vector<uint8_t>> DecompressionStreamDecoder::decompress(const uint8_t* input, const size_t inputLength)
+{
+    size_t index = 0, allocateSize = 16384, totalSize = 0;
```
Ditto: Magic number.

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -32,15 +32,121 @@
+            shouldDecompress = false;
+        else {
+            index++;
+            if (allocateSize < 1073741824) // 1GB
```
Ditto: Magic number...

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -32,15 +32,121 @@
+
+    Vector<uint8_t> output;
+    output.reserveCapacity(totalSize);
+    for (size_t i = 0; i <= index; i++)
```
Ditto modern C++ for loop

brentfulgham commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -41,17 +41,31 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
-    const Formats::CompressionFormat m_format;
+    
+    bool initailized;
+    bool finish;
```
Ditto use modern C++ initializers here.

brentfulgham reviewed (changes requested):
Looks good -- I had a few stylistic and maintenance fixes I'd like to see before landing, but really nice job!!!!

brentfulgham reviewed (approved):
Looks good!

stwrt committed:
Compression Streams
https://bugs.webkit.org/show_bug.cgi?id=242665

Reviewed by Brent Fulgham.

This patch implements the Compression Streams API.
Currently we are passing all WPT test cases (except for an initialization one where
we are throwing a typeError instead of a standard Error).

This implementation uses the compression and decompression functions from the Zlib library.

For compression I attempt to perform the compression in one shot by allocating enough memory up front per chunk.
For decompression I am using a scaling allocation approach. We first allocate 1 page (16KB) of memory, then two, then 4 and so on...
I added a hard limit of 1GB per allocation to make sure I don't blow away all the device's memory.
For every chunk, I take all these allocations and merge them into one array.

* LayoutTests/imported/w3c/web-platform-tests/compression/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.js: Added.
(get value):
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-bad-chunks.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.js: Added.
(async compressChunkList):
(const.chunkList.of.chunkLists.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.js: Added.
(makeExpectedChunk):
(async compressMultipleChunks):
(numberOfChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-multiple-chunks.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.js: Added.
(async compressArrayBuffer):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-output-length.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.js: Added.
(async compressArrayBuffer):
(test):
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-stream.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-with-detach.tentative.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-with-detach.tentative.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-with-detach.tentative.window.js: Added.
(createLargeRandomInput):
(decompress):
(promise_test.async const):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.js: Added.
(get value):
(return.new.Uint8Array):
(value.new.Uint8Array):
(const.chunk.of.badChunks.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-bad-chunks.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.js: Added.
(value.new.Int8Array.new.Uint8Array):
(value.new.Uint8Array):
(value.new.Uint8ClampedArray):
(value.new.Int16Array):
(value.new.Uint16Array):
(value.new.Int32Array):
(value.new.Uint32Array):
(value.new.Float32Array):
(value.new.Float64Array):
(const.chunk.of.bufferSourceChunksForDeflate.promise_test.async t):
(const.chunk.of.bufferSourceChunksForGzip.promise_test.async t):
(const.chunk.of.bufferSourceChunksForDeflateRaw.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-buffersource.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.js: Added.
(test.t.new.DecompressionStream.toString):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-constructor-error.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.js: Added.
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-correct-input.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any.js: Added.
(async tryDecompress):
(corruptInput):
(of.expectations.promise_test.async const):
(of.expectations.promise_test):
(of.expectations.of.fields.of.cases.promise_test.async const):
(of.expectations.of.fields.of.cases.promise_test):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-corrupt-input.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.js: Added.
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-empty-input.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.js: Added.
(async decompressArrayBuffer):
(chunkSize.promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-split-chunk.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.js: Added.
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-uint8array-output.tentative.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-with-detach.tentative.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-with-detach.tentative.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/decompression-with-detach.tentative.window.js: Added.
(async createLargeCompressedInput):
(promise_test.async const):
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness-shadowrealm.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness-shadowrealm.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness-shadowrealm.window.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/idlharness.https.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/resources/concatenate-stream.js: Added.
(async concatenateStream):
* LayoutTests/imported/w3c/web-platform-tests/compression/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/third_party/pako/LICENSE: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/third_party/pako/README: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/third_party/pako/pako_inflate.min.js: Added.
(l):
(return.r):
(i.assign):
(i.shrinkBuf):
(a.arraySet):
(a.flattenChunks):
(r.arraySet):
(r.flattenChunks):
(i.setTyped):
(1):
(d):
(i.string2buf):
(i.buf2binstring):
(i.binstring2buf):
(i.buf2string):
(i.utf8border):
(2):
(t.exports):
(3):
(4):
(s):
(5):
(6):
(7):
(L):
(r):
(o):
(f):
(H):
(j):
(i.inflateInit):
(i.inflate):
(i.inflateEnd):
(i.inflateGetHeader):
(i.inflateSetDictionary):
(8):
(9):
(10):
(11):
(o.prototype.push):
(o.prototype.onData):
(o.prototype.onEnd):
(i.inflateRaw):
(string_appeared_here):
* LayoutTests/imported/w3c/web-platform-tests/compression/third_party/pako/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/w3c-import.log: Added.
* Source/WebCore/Modules/compression/CompressionStream.js:
(initializeCompressionStream):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
(WebCore::CompressionStreamEncoder::encode):
(WebCore::CompressionStreamEncoder::flush):
(WebCore::CompressionStreamEncoder::initialize):
(WebCore::CompressionStreamEncoder::compress):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
(WebCore::CompressionStreamEncoder::~CompressionStreamEncoder):
(WebCore::CompressionStreamEncoder::CompressionStreamEncoder):
* Source/WebCore/Modules/compression/DecompressionStream.js:
(initializeDecompressionStream):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::decode):
(WebCore::DecompressionStreamDecoder::flush):
(WebCore::DecompressionStreamDecoder::initialize):
(WebCore::DecompressionStreamDecoder::decompress):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
(WebCore::DecompressionStreamDecoder::~DecompressionStreamDecoder):
(WebCore::DecompressionStreamDecoder::DecompressionStreamDecoder):
* Source/WebCore/Modules/compression/Formats.h:

Canonical link: https://commits.webkit.org/253093@main
